exchanges/okx/okx_wrapper.go

```
@@ -0,0 +1,1218 @@
+			goto checkErrorAndContinue
+		}
+		instsb, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "ETH-USD",
+		})
+		if er != nil {
+			goto checkErrorAndContinue
+		}
+		instsc, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "SOL-USD",
+		})
+		insts = append(insts, instsb...)
+		insts = append(insts, instsc...)
+	case asset.Margin:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeMargin,
+		})
+	}
+checkErrorAndContinue:
+	if er != nil || len(insts) == 0 {
+		return pairs, er
+	}
+	for x := range insts {
+		var pair string
+		switch insts[x].InstrumentType {
+		case asset.Spot:
+			pair = insts[x].BaseCurrency + format.Delimiter + insts[x].QuoteCurrency
+		case asset.Futures, asset.PerpetualSwap, asset.Option:
+			currency, err := currency.NewPairFromString(insts[x].Underlying)
+			if err != nil {
+				continue
+			}
+			pair = currency.Base.String() + format.Delimiter + currency.Quote.String()
+		}
+		pairs = append(pairs, pair)
+	}
+	return pairs, nil
+}
+
+// UpdateTradablePairs updates the exchanges available pairs and stores them in the exchanges config
+func (ok *Okx) UpdateTradablePairs(ctx context.Context, forceUpdate bool) error {
+	assetTypes := ok.GetAssetTypes(false)
+	for i := range assetTypes {
+		if !(assetTypes[i] == asset.Spot || assetTypes[i] == asset.PerpetualSwap || assetTypes[i] == asset.Option || assetTypes[i] == asset.Futures || assetTypes[i] == asset.Margin) {
+			return errInvalidInstrumentType
+		}
+		p, err := ok.FetchTradablePairs(ctx, assetTypes[i])
+		if err != nil {
+			return err
+		}
+		selectedPairs := []string{}
+		pairsMap := map[string]int{}
+		for i := range p {
+			if p[i] == "" {
+				continue
+			}
+			count, ok := pairsMap[p[i]]
+			if !ok || count == 0 {
+				pairsMap[p[i]] = 1
+				selectedPairs = append(selectedPairs, p[i])
+			} else {
+				continue
```

if there is a duplication we should return an error.